File: fast_barcode_scanner/ios/Classes/types/ScannerError.swift
import Flutter

extension Error {
    var flutterError: FlutterError {
        if let error = self as? ScannerError {
            return error.flutterError
        } else {
            return FlutterError(code: "UNEXPECTED_ERROR",
                                message: "Unknown error occured.",
                                details: nil)
        }
    }
}

enum ScannerError: Error {
    case notInitialized
    case alreadyInitialized
    case notRunning
    case alreadyRunning
    case noInputDeviceForConfig(ScannerConfiguration)
    case cameraNotSuitable(Resolution, Framerate)
    case unauthorized
    case configurationError(String)
    case invalidArguments(Any?)
    case invalidCodeType(String)
    case loadingFailed
    case analysisFailed(Error)
    case minimumTarget
    case loadingURLFailed(String)
    case loadingDataFailed

    var flutterError: FlutterError {
        switch self {
        case .notInitialized:
            return FlutterError(code: "NOT_INITIALIZED",
                                message: "Camera has not been initialized",
                                details: nil)
        case .alreadyInitialized:
            return FlutterError(code: "ALREADY_INITIALIZED",
                                message: "Camera is already initialized",
                                details: nil)
        case .notRunning:
            return FlutterError(code: "NOT_RUNNING",
                                message: "Camera is not running",
                                details: nil)
        case .alreadyRunning:
            return FlutterError(code: "ALREADY_RUNNING",
                                message: "Camera is already running",
                                details: nil)
        case .cameraNotSuitable(let res, let fps):
            return FlutterError(code: "CAMERA_NOT_SUITABLE",
                                message: """
                                    The camera does not support the requested resolution (\(res)) \
                                    and framerate (\(fps)) combination
                                    """,
                                details: "Try to lower your settings")
        case .configurationError(let error):
            return FlutterError(code: "CONFIGURATION_FAILED",
                                message: "Configuration failed (\(error))",
                                details: nil)
        case .noInputDeviceForConfig(let config):
            return FlutterError(code: "NO_INPUT_DEVICE",
                                message: "No input device found for configuration (\(config)",
                                details: "Are you using a simulator?")
        case .unauthorized:
            return FlutterError(code: "UNAUTHORIZED",
                                message: "The application is not authorized to use the camera device",
                                details: nil)
        case .invalidArguments(let args):
            return FlutterError(code: "INVALID_ARGUMENT",
                                message: "Invalid arguments provided (\(String(describing: args)))",
                                details: nil)
        case .invalidCodeType(let type):
            return FlutterError(code: "INVALID_CODE",
                                message: "Invalid code type \(type)",
                                details: nil)
        case .loadingFailed:
            return FlutterError(code: "LOADING_FAILED",
                                message: "Could not load asset",
                                details: nil)
        case .analysisFailed(let error):
            return FlutterError(code: "ANALYSIS_FAILED",
                                message: "Could not analyse asset",
                                details: error.localizedDescription)
        case .minimumTarget:
            return FlutterError(code: "MINIMUM_TARGET",
                                message: "iOS 11 or higher required",
                                details: nil)
        case .loadingURLFailed(let url):
            return FlutterError(code: "LOADING_URL_FAILED",
                                message: "Loading image from url failed",
                                details: url)
        case .loadingDataFailed:
            return FlutterError(code: "LOADING_DATA_FAILED",
                                message: "Loading image from data failed",
                                details: nil)
        }
    }
}

File: fast_barcode_scanner/android/src/main/kotlin/com/jhoogstraat/fast_barcode_scanner/types/ScannerError.kt
package com.jhoogstraat.fast_barcode_scanner.types

import io.flutter.plugin.common.MethodChannel.Result
import java.io.IOException
import java.util.*
import kotlin.collections.HashMap

sealed class ScannerError : Throwable() {
    class NotInitialized : ScannerError()
    class NotRunning : ScannerError()
    class AlreadyRunning : ScannerError()
    class NoInputDeviceForConfig(val configuration: ScannerConfiguration) : ScannerError()
    class CameraNotSuitable(val resolution: Resolution, val framerate: Framerate) : ScannerError()
    class Unauthorized : ScannerError()
    class ConfigurationError(val error: Exception) : ScannerError()
    class InvalidArguments(val args: HashMap<String, Any>) : ScannerError()
    class InvalidCodeType(val type: String) : ScannerError()
    class LoadingFailed(val error: IOException) : ScannerError()
    class AnalysisFailed(val error: Exception) : ScannerError()

    /* Android specific */
    class ActivityNotConnected : ScannerError()

    fun throwFlutterError(result: Result) {
        return when(this) {
            is NotInitialized -> result.error("NOT_INITIALIZED", "Camera has not been initialized", null)
            is NotRunning -> result.error("NOT_RUNNING", "Camera is not running", null)
            is AlreadyRunning -> result.error("ALREADY_RUNNING", "Camera is already running", null)
            is CameraNotSuitable -> result.error("CAMERA_NOT_SUITABLE", "The camera does not support the requested resolution and framerate combination", "$resolution $framerate")
            is ConfigurationError -> result.error("CONFIGURATION_FAILED", "The configuration could not be applied", error.localizedMessage)
            is NoInputDeviceForConfig -> result.error("NO_INPUT_DEVICE", "No input device found for configuration. Are you using a simulator?", "$configuration")
            is Unauthorized -> result.error("UNAUTHORIZED", "The application is not authorized to use the camera device", null)
            is InvalidArguments -> result.error("INVALID_ARGUMENT", "Invalid arguments provided", args)
            is InvalidCodeType -> result.error("INVALID_CODE", "Invalid code type", type)
            is ActivityNotConnected -> result.error("NO_ACTIVITY", "No activity is connected", null)
            is LoadingFailed -> result.error("LOADING_FAILED", "Could not load asset", error.localizedMessage)
            is AnalysisFailed -> result.error("ANALYSIS_FAILED", "Could not analyse asset", error.localizedMessage)
        }
    }
}